File: 2024/May/4-2 2441 Largest Positive Integer That Exists With Its Negative/Set Approach/LargestMatch.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var findMaxK = function(nums) {
    let positives = new Set();
    let negatives = new Set();
    let num = -1;

    for(let i of nums) {
        if(i > 0) {
            if(negatives.has(i)) {
                if(i > num) {
                    num = i;
                }
            }
            else {
                positives.add(i);
            }
        }
        else {
            let pos = Math.abs(i);
            if(positives.has(pos)) {
                if(pos > num) {
                    num = pos;
                }
            }
            else {
                negatives.add(pos);
            }
        }
    }

    return num;
};

File: 2024/May/5-2 2441 Largest Positive Integer That Exists With Its Negative/Queue Approach/LargestMatch.java
class Solution {
    public int findMaxK(int[] nums) {
        Queue<Integer> negatives = new PriorityQueue<Integer>((a,b) -> b - a);
        Queue<Integer> positives = new PriorityQueue<Integer>((a,b) -> b - a);
        
        for(int i : nums) {
            if(i < 0) {
                negatives.offer(Math.abs(i));
            }
            else {
                positives.offer(Math.abs(i));
            }
        }

        while(!positives.isEmpty() && !negatives.isEmpty()) {
            if(Integer.compare(positives.peek(), negatives.peek()) == 0) {
                return positives.poll();
            }
            
            if(Integer.compare(positives.peek(), negatives.peek()) > 0) {
                positives.poll();
            }
            else {
                negatives.poll();
            }
        }

        return -1;
    }
}